### ISSUE TITLE
Building directories via ActivityPub

### ISSUE THREAD

**Issue by gabek** [owncast/owncast#1143]
# Goal

## Support self-hosted, independent mini-directories of Owncast instances or "Featured streams"

Different people will have different ideas of what this means.  One person who says they want to have a directory means they want to download a complete product that they can run.  Another may simply want the ability to put together a directory via APIs.

The focus will be creating the building blocks via ActivityPub so anybody who follows an Owncast instance will get the necessary information needed to generate a directory around it. Including ourselves.

Eventually we migrate the existing https://owncast.directory to use this new ActivityPub backbone.  And once that's done maybe we can release that as an example directory implementation. But that's probably a bit down the road.

## Feature

For this actual feature we will build new "featured streams" functionality where an Owncast server will follow other Owncast servers via ActivityPub and exchange stream information via that method. That also means that if anybody else wants to build their own directories, they can also do so over ActivityPub as the mechanism.

- Show a list of "Featured Streams" in the UI that a user can browse, and see if they're live.
- If the current server you're on is not live, it will recommend a featured server that is live that you should visit instead.

We *could* also allow an Owncast server to follow non-Owncast actors on the Fediverse and link to them. And there's no huge downside to that except it would mix up live streams and people, and featuring people doesn't make a ton of sense. It would be difficult to show context around why they're there in the UI.

---

### Go Live

Create Note
- This already exists and is already sent to the Fediverse.
- People can already tie into this.
- But new metadata will be added.
- This is a note and not a `Video` object because it's meant to be consumed by actual people on microblogging platforms, not just other systems. This is the "I've gone live!" message. Anything other than a note on microblogging platforms are attempted to be handled as best as possible, or thrown away completely. So this is our compromise.

### Stream ended

[Leave Video](https://www.w3.org/TR/activitystreams-vocabulary/#dfn-leave)

When the stream ends this activity is sent out.

### Stream ping

[Offer Video ](https://www.w3.org/TR/activitystreams-vocabulary/#dfn-offer)

Let all followers know that the stream is still live so directories can implement "last seen at" logic.

---

# What's required (a work in progress list)

- [x] Owncast must be able to send out Follow requests to actors that it wishes to feature.
- [x] Owncast to send out activities around state (stream start, stream ping, stream stop) to those who are following.
- [ ] Build a new API so an Owncast server can tell any directory that it wants to be listed, therefore the directory should follow the instance. (Server B asks Server A to list it. Server A has to approve). This would be less for the "Featured streams" and more for larger directories when an Owncast server wants to be listed on it.
- [x] Admin interface to show a list of Owncast servers that are listed so individual ones can be removed (therefore being removed from that directory).
- [x] Owncast server A can add Owncast server B to its list of featured live streams. (Follow request)
- [ ] Owncast server B would have to approve featuring. (Follow approval)
- [ ] When Owncast server B goes live, it shows Owncast server B as live on Owncast server B's page. (Our existing "go live" message that gets sent out, but with a new property we'll populate so Owncast knows what the Activity actually is)

**Comment by GuiguiSurLinux** [comment]
Hello! I'm very interested in independent directories. I see a lot of these WIP are already done, if I can get a sense of what's left I would be more than happy to learn Go and submit an MR (I'm a Python dev but I'm told the paradigms are similar).

For context, I'm a recent Owncast convert, thinking of building my own directory to encourage people to make the switch as well (I've had a few people tell me they'd love to, if only there were more of a local community). There is a strong anti-Amazon sentiment here currently that I think is making people reconsider their streaming platform.

In technical terms, my initial idea was simply to show all streamers with a specific tag in the current directory. I'd personally rather depend on a "main" directory like this, since people who want to be in a directory are already there, but I understand this could put undue pressure on it when compared with independent instances.

In any case, let me know if and how I can help on this issue.

PS many thanks to all contributors for this amazing piece of software.

**Comment by gabek** [comment]
Hi there, thanks for the kind words!

None of the work required for this functionality has even started yet, so there's a lot that needs to happen.  If you look at the [roadmap](https://owncast.online/roadmap/) you'll see "Federated instance follows" as a roadmap item, and that's what needs to happen before we can start talking about independent directories. It's probably a year or two out, honestly, as it's not just one change that has to happen, it's an entire new federated feature set. And that's not scheduled to be worked on until other things are completed first.

In the meantime, you can still build your own directory using Owncast webhooks. Build a site that accepts webhooks for stream start and stop, and any Owncast instance that wants to show up on your directory can add your site as a webhook destination for start and stop webhooks. Then your site can do whatever it wants with regard to showing or hiding instances based on any criteria you'd like.

**Comment by gabek** [comment]
I've moved things around, and I'm making this the next major feature. `v0.3.0` will have directory support powered by ActivityPub. It will be surfaced in Owncast as a way to feature other streams within your Owncast server, but big picture it will allow anybody to build directories using just ActivityPub.
